source/extensions/filters/common/ratelimit/ratelimit_impl.cc

```diff
@@ -0,0 +1,145 @@
+#include "extensions/filters/common/ratelimit/ratelimit_impl.h"
+
+#include <chrono>
+#include <cstdint>
+#include <string>
+#include <vector>
+
+#include "envoy/api/v2/ratelimit/ratelimit.pb.h"
+#include "envoy/registry/registry.h"
+#include "envoy/singleton/manager.h"
+#include "envoy/stats/scope.h"
+
+#include "common/common/assert.h"
+#include "common/http/header_map_impl.h"
+#include "common/http/headers.h"
+
+#include "server/configuration_impl.h"
+
+namespace Envoy {
+namespace Extensions {
+namespace Filters {
+namespace Common {
+namespace RateLimit {
+
+// Singleton registration via macro defined in envoy/singleton/manager.h
```

Comment:
Currently also it does not allow registering twice - Here I am registering client. I can move that to registration file.

Reply:
On a second thought, I think it is better to leave the singleton registration of client here it self because, once we fully migrate to the new model and remove rate limit service from bootstrap, we can just remove the whole registration file. I will documentation to indicate the same.

Reply:
Sorry I still do not understand why we have to singleton register the config at all. I think we should only be singleton registering the client:
1) Bootstrap -> Create client (or null) and register it
2) Filter, get singleton client

In the future, we can add client config, and then in 2) we can also attempt to register, and then we can remove 1. What am I missing?

Reply:
We can do what you said above. But we can only register the client once. So if we add the client config, we can not register from multiple places (bootstrap and filter). We will have to remove at one place - we are not sure at which place we can remove when both co-exist during the deprecation period.  The advantage of registering config is we can validate during the deprecation period that one supplied in bootstrap and client config is same and register client at only one place (that is from filter code as I am doing today). If the validation fails, we can fail fast. If we do not store config in singleton, I do not know how it we can work during the deprecation period - if core code already registers client and we can not register again with the client config in filters.
So leaving the registration of client in the filter code will allow us to validate and register it with right config (either bootstrap or client)?
Does it make sense?

Reply:
The point of the singleton manager is that you can attempt to register from multiple places, but it will only get registered once. So basically you have common code that looks like this:
1) See if singleton exists
2) If so, check if config matches, if not fail config
3) If not, create singleton

This should work fine.

Reply:
When I attempt to register the client from filter code with filter config, how can I get the bootstrap config to validate in 2 above? Are you suggesting to hold the config in the GrpcClientImpl (global rate limit client)? I am probably missing that part.

Reply:
Yes, I would just store the proto config used to create the client, and then compare it.

Reply:
Ok. Now it make sense. I will make that change.

Reply:
I moved the whole client creation/registration to ratelimit_registration so that singletons are maintained at single place.

Reply:
OK that's fine. I think we can clean this up once we deprecate the bootstrap config, and then we can have a single client for the entire server which would be nice.